### ISSUE TITLE
Iteration Plan for January 2021

### ISSUE THREAD

**Issue by kieferrm** [microsoft/vscode#112419]
This plan captures our work in late **December** and **January**. This is a 6-week iteration. We will ship in **early February**.

> **Note**: Our last Insiders build of 2020 will go out on December 22. The first Insiders build of 2021 will go out on January 5. Our team will have minimal presence between December 19 and January 3. Expect slow or no responses in this time frame. Thank you for your support and contributions in 2020! Happy Holidays to all of you.

## Endgame
- **January 25, 2021**: Endgame begins
- **January 29, 2021**: Endgame done

The endgame details for this iteration are tracked in https://github.com/microsoft/vscode/issues/114896.

## Plan Items
Below is a summary of the top level plan items.

Legend of annotations:

| Mark                 | Description                                       |
| -------------------- | ------------------------------------------------- |
| :runner:             | work in progress                                  |
| :hand:               | blocked task                                      |
| :muscle:             | stretch goal for this iteration                   |
| :red_circle:         | missing issue reference                           |
| :large_blue_circle:  | more investigation required to remove uncertainty |
| :black_circle:       | under discussion within the team                  |
| :black_large_square: | a large work item, larger than one iteration      |

### Accessibility
- [x] Accessibility issues, see [query](https://github.com/Microsoft/vscode/issues?q=is%3Aopen+is%3Aissue+label%3Aaccessibility+milestone%3A%22January+2021%22) **team**
- [ ] :runner: Assess notebook accessibility https://github.com/microsoft/vscode/issues/111255 @isidorn @rebornix @roblourens

### Workbench
- [ ] :runner: Explore the concept of Trusted Workspaces https://github.com/microsoft/vscode/issues/106488 @lszomoru @sbatten
- [x] Investigate predefined file filters in search https://github.com/microsoft/vscode/issues/106790 @jacksonkearl
- [ ] :runner: Improve the welcome pages https://github.com/microsoft/vscode/issues/106715 @misolori @jacksonkearl @digitarald
- [ ] :runner: :muscle: Explore support for adding/organizing imports on Paste https://github.com/microsoft/vscode/issues/30066 @mjbvz
- [x] Improve emmet extension perf https://github.com/microsoft/vscode/issues/71996 @rzhao271
- [x] Migrate emmet dependency to latest https://github.com/microsoft/vscode/issues/110697 @rzhao271
- [x] Show file decorations in editor tabs https://github.com/microsoft/vscode/issues/49382 @jrieken
- [x] Improve undo/redo from explorer when an extension contributes edits https://github.com/microsoft/vscode/issues/111867 @jrieken @isidorn
- [x] Allow tabs to wrap to multi-line https://github.com/microsoft/vscode/issues/70413 @bpasero
- [x] Support to customize the product icons https://github.com/microsoft/vscode/issues/92791 @aeschli
- [x] Provide hovers in comments views https://github.com/microsoft/vscode/issues/108396 @rmacfarlane
- [x] Revisit how activity bar items and status bar items can be hidden by the user https://github.com/microsoft/vscode/issues/113757 @bpasero @sbatten
- [ ] :runner: Explore consolidation of managing "Don't Show Again" in notifications https://github.com/microsoft/vscode/issues/24815 @meganrogge
- [x] Utilize the GH login for the Issue Reporter https://github.com/microsoft/vscode/issues/9516 @rmacfarlane
- [x] Introduce a setting to switch between the Search editor and the Search view https://github.com/microsoft/vscode/issues/112588 @jacksonkearl

### Code Editor

- [x] Inline Hint API and UI proposal, https://github.com/microsoft/vscode/issues/113276 @jrieken

### Notebook Editor
- [x] Identify and fill API gaps for LiveShare https://github.com/microsoft/vscode/issues/102503 @rebornix
- [x] Explore notebook keymap support https://github.com/microsoft/vscode/issues/106679 @rebornix @roblourens
- [ ] :runner: Explore extensible markdown rendering for notebooks https://github.com/microsoft/vscode/issues/106701 @mjbvz
- [x] Improve notebook diff editor scrolling performance https://github.com/microsoft/vscode/issues/109680 @rebornix
- [x] Explore output renderer support in notebook diff editors https://github.com/microsoft/vscode/issues?q=is%3Aopen+label%3Anotebook+diff @rebornix
- [x] Explore options to support free-floating output renderer elements https://github.com/microsoft/vscode/issues/108947 @rebornix
- [ ] :runner: Explore options to allow an output to go into full-screen mode https://github.com/microsoft/vscode/issues/110766 @rebornix @deepak1556
- [x] Provide Outline View support for Notebooks https://github.com/microsoft/vscode/issues/98173 @jrieken
- [ ] Improve focus handling in notebooks https://github.com/microsoft/vscode/issues/113246 @misolori
- [ ] Improve discoverability of cell folding https://github.com/microsoft/vscode/issues/113248 @misolori @rebornix
- [ ] :runner: Explore cell multi-select https://github.com/microsoft/vscode/issues/114156 @rebornix @misolori

#### LSP /LSIF
- [ ] :runner: LSP - investigate supporting a pull model for diagnostics (diagnostics are currently delivered using a push model) https://github.com/microsoft/language-server-protocol/issues/737 @dbaeumer
- [x] Enable showing completion item detail if available for all completion items https://github.com/microsoft/vscode/issues/39441 @dbaeumer
- [x] Implement an LSIF index validator https://github.com/microsoft/lsif-node/issues/119 @dbaeumer

#### JavaScript/TypeScript
- [X] Enable TS 4.2 https://github.com/microsoft/vscode/issues/115255 @mjbvz

### Debug
- [x] Update all node debug documentation/tutorials to use `debug-js` https://github.com/microsoft/vscode-docs/issues/4123 @connor4312
- [x] Eliminate "preview" term from debuggers (Node, Chrome, ExtDev) https://github.com/microsoft/vscode/issues/112139 @connor4312
- [ ] :runner: Extension API for "Inline Values" https://github.com/microsoft/vscode/issues/105690 @weinand
- [ ] Improve the "breakpoint hit" UI for Web UI https://github.com/microsoft/vscode/issues/100757 @weinand @isidorn
- [x] Support to edit the condition expression for function breakpoints https://github.com/microsoft/vscode/issues/3646 @isidorn
- [x] Support to start multiple debug sessions from a single launch config https://github.com/microsoft/vscode/issues/110226 @isidorn
- [ ] :runner: :muscle: Ability to drag and drop breakpoints https://github.com/microsoft/vscode/issues/88227 @isidorn

#### DAP
- [ ] :runner: Continue discussion: Support object IDs https://github.com/microsoft/debug-adapter-protocol/issues/134 @weinand
- [ ] :runner: Investigate: Control escaping/quoting arguments for the `runInTerminal` request https://github.com/microsoft/debug-adapter-protocol/issues/146 @weinand
- [ ] Explore support for large number of threads https://github.com/microsoft/debug-adapter-protocol/issues/169 @weinand @connor4312

### Testing
- [ ] :runner: Explore how to improve testing support https://github.com/microsoft/vscode/issues/107467 @connor4312 @sandy081

### Terminal
- [x] Restore terminal UI state https://github.com/microsoft/vscode/issues/109244 @roblourens @meganrogge
- [x] Provide flow control in the terminal to avoid starvation in the case of continuous output https://github.com/microsoft/vscode/issues/113827 @tyriar @meganrogge
- [x] Investigate memory leak in Terminal https://github.com/microsoft/vscode/issues/111562 @meganrogge @tyriar
- [x] Improve predictive echo for PowerShell https://github.com/microsoft/vscode/issues/113853 https://github.com/microsoft/vscode/issues/113857 @TylerLeonhardt
- [x] Support Powershell from the Windows Store https://github.com/microsoft/vscode/issues/110828 @TylerLeonhardt

### Source Control
- [x] Warn when pushing to a rebased branch https://github.com/microsoft/vscode/issues/110881 @eamodio
- [ ] :muscle: Guide user to push to a remote after creating a commit https://github.com/microsoft/vscode/issues/110882 @eamodio

### API
- API finalization: [query](https://github.com/Microsoft/vscode/issues?q=is%3Aissue+is%3Aopen+milestone%3A%22January+2021%22+label%3A%22api-finalization%22)
- API proposals: [query](https://github.com/Microsoft/vscode/issues?q=is%3Aissue+is%3Aopen+milestone%3A%22January+2021%22+label%3A%22api-proposal%22)

### Extensions

- [x] Provide design and UX guidelines for extension authors https://github.com/microsoft/vscode/issues/104006 @misolori

### Extension Contributions
- [x] Improvements to the Remote extensions https://github.com/microsoft/vscode-remote-release/issues/4184 **team**
- [x] Improvements to the GitHub PR extension https://github.com/microsoft/vscode-pull-request-github/issues/2293 @rmacfarlane @eamodio
- [x] Continue improving and adding functionality to REST Book https://github.com/tanhakabir/rest-book/issues/30 @tanhakabir

### Install
- [x] Explore installing a WSL extension out of the box on Windows https://github.com/microsoft/vscode/issues/106759 @joaomoreno @aeschli
- [x] :red_circle: Explore to show macOS ARM downloads side by side with the Intel version @deepak1556 @gregvanl

### Engineering
- [x] Continuous build optimizations (run tests in parallel) @lszomoru @joaomoreno @alexdima
- [ ] :runner: :red_circle: Investigate VS Code build optimization: separate the building of the core from the building of the extensions @joaomoreno
- [x] Publish Linux ARM deb, rpm and snap to repositories https://github.com/Microsoft/vscode/issues/106473 @joaomoreno
- [x] Make VSCode [Trusted Types](https://w3c.github.io/webappsec-trusted-types/dist/spec/) compliant https://github.com/Microsoft/vscode/issues/103699 @jrieken and **team**
- [x] Stabilize Apple Silicon builds https://github.com/microsoft/vscode/issues/106770 @deepak1556
- [x] MacOS universal build https://github.com/microsoft/vscode/issues/113851 @deepak1556
- [ ] :runner: electron 11 stabilization https://github.com/microsoft/vscode/issues?q=is%3Aissue+label%3Aelectron-11-update+is%3Aopen @deepak1556 @rzhao271
- [ ] :runner: Terminal lag on Big Sur https://github.com/microsoft/vscode/issues/105446 @deepak1556
- [ ] :muscle: Scroll performance on Windows https://github.com/microsoft/vscode/issues/107016 @deepak1556
- [ ] :runner: Move away from `get*Actions` in Views and ViewContainers and use Menu instead https://github.com/microsoft/vscode/issues/92038 @sandy081 **team**
- [ ] :red_circle: Adopt `vscode.open` and `vscode.diff` in SCM, Timeline, GHPRI extension
  - [ ] Timeline @eamodio
  - [ ] :runner: GHPRI extension @alexr00
  - [x] Custom Views @alexr00
- [ ] :red_circle: Investigate using component governance support for third party notices generation @lszomoru

#### Node.js-free, sandboxed renderer
- [ ] :runner: Backlog https://github.com/microsoft/vscode/issues/92164 @bpasero @deepak1556

**Comment by Igorgro** [comment]
When new Native Notebook Editor will be available in stable version?

**Comment by axetroy** [comment]
Digression

I think this statistic (https://vscode-github-stats.netlify.app/) is great.

This is for the vscode team to better manage and maintain the issue.

But I wonder if there are statistics for users, which will count the features that users most want and the bugs that they most want to fix.

| Most anticipated features                                                 | Most eager to fix                                         |
| ------------------------------------------------------------------------- | --------------------------------------------------------- |
| [Allow to hide editor title section entirely (tabs, breadcrumbs)](#33607) | [Slow reponse to scroll and typing](#107016)              |
| [Allow multiple views to show at the same time](#26777)                   | [VS Code was lost on shutdown with pending update](52855) |
| ...                                                                       | ...                                                       |
| ...                                                                       | ...                                                       |
| ...                                                                       | ...                                                       |
| ...                                                                       | ...                                                       |
| ...                                                                       | ...                                                       |
| ...                                                                       | ...                                                       |
| ...                                                                       | ...                                                       |
| ...                                                                       | ...                                                       |
| ...                                                                       | ...                                                       |
| ...                                                                       | ...                                                       |

in technology, it’s not difficult to count this information. (From the issue, you can calculate the number of likes, creating time, number of participants in the discussion, and even the number of vscode team members participating in the discussion)

Because some issues are 2 years or even 3 years ago, and the attention is not a lot.

Hope such statistics can help the vscode team to understand what users want most.

thanks for your great work.

**Comment by GitMensch** [comment]
> Because some issues are 2 years or even 3 years ago, and the attention is a lot.

You mean things like #2809... the question is: do the vscode team want these statistics?

**Comment by xianghongai** [comment]
Add this feature?  😘

![VSCode-Foldable-in-Markdown-1](https://user-images.githubusercontent.com/2075006/102858738-ba205980-4465-11eb-8f63-306e6440a077.png)

![VSCode-Foldable-in-Markdown-2](https://user-images.githubusercontent.com/2075006/102858741-bb518680-4465-11eb-916b-421bca55edc9.png)

**Comment by GitMensch** [comment]
I guess that's not a wish list here, but #25887 (extend diff editor with merge fetures) has no downvote and quite a big community-request-level:
![grafik](https://user-images.githubusercontent.com/6699539/102910490-f1e5cc00-447a-11eb-8ae8-ddfe2604f803.png)

**Comment by AnrDaemon** [comment]
> But I wonder if there are statistics for users, which will count the features that users most want and the bugs that they most want to fix.

When there are 4k++ open issues, it's rather hard to find the one relevant to your problem. Which makes it problematic to amass any representative number of responses to any single problem.

**Comment by AnrDaemon** [comment]
> I think this statistic (https://vscode-github-stats.netlify.app/) is great.

I think statistics of a moving average of opened and closed issues per 10 days would be more representative than absolute numbers.

**Comment by sanket-bhalerao** [comment]
Having that update notification on the insider's build after so long feels sooo good 🎉.

**Comment by lucastrvsn** [comment]
Any chance of [https://github.com/microsoft/vscode/issues/5280](https://github.com/microsoft/vscode/issues/5280) to be included on this roadmap? This minor feature will allow keybinding extensions to do so much! Thanks.

**Comment by KamasamaK** [comment]
Any plans to update the [yearly roadmap](https://github.com/microsoft/vscode/wiki/Roadmap) for 2021 in this iteration?

**Comment by demonguy** [comment]
i still wonder if we're going to have 3-way merge diff editor

**Comment by PathToLife** [comment]
@lucastrvsn 
#5280 pr made https://github.com/microsoft/vscode/pull/115190

**Comment by objectivecosta** [comment]
Is a macOS ARM version coming this release? I see the bullet point, but no issue #.

**Comment by LiuYiwai** [comment]
> Is a macOS ARM version coming this release? I see the bullet point, but no issue #.

I think yes, I see the relevant PR in homebrew-cask.

**Comment by GitMensch** [comment]
> January 29, 2021: Endgame done

vs.

> The endgame details for this iteration are tracked in TBD.

Can you share an updated time for this? We all know how burdened those times are, but it likely helps some people to know more about the expected new date. If there's still much to do then it is completely fine to make it a December->January->February iteration plan, the only thing that is bad is to stay on "done end of January" when it did not started yet and is February.

**Comment by Tyriar** [comment]
@GitMensch I filled in the link: https://github.com/microsoft/vscode/issues/114896, this just like our other releases have work done during the month it was named and released early the following month. Expect the release later this week, but it's possible it could get pushed to Monday.
